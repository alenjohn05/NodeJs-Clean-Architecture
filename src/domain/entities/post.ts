export class Post {
  public id?: number;
  public createdAt: Date;
  public published: boolean;
  public title: string;

  constructor(post: { id?: number; createdAt: Date; published: boolean; title: string }) {
    this.id = post.id;
    this.createdAt = post.createdAt;
    this.published = post.published;
    this.title = post.title;
  }
}
