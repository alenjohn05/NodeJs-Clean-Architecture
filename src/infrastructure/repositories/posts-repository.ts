import { Post as PostModel } from '@prisma/client';
import { IPostsRepository } from '@application/common/interfaces';
import { Dependencies } from '@infrastructure/di';
import { Post } from '@domain/entities';

export function makePostsRepository({ db }: Pick<Dependencies, 'db'>): IPostsRepository {
  return {
    async create({ post }) {
      const { id } = await db.post.create({
        data: {
          createdAt: post.createdAt,
          published: post.published,
          title: post.title,
        },
      });

      return {
        id,
      };
    },
    async delete({ id }) {
      await db.post.delete({ where: { id } });
    },
    async getById({ id }) {
      const post = await db.post.findFirst({ where: { id } });

      if (!post) {
        return null;
      }

      return toEntity(post);
    },
    async list({ pageNumber, pageSize }) {
      const count = await db.post.count();

      const posts = await db.post.findMany({
        skip: (pageNumber - 1) * pageSize,
        take: pageSize,
      });

      return {
        count,
        posts: posts.map(toEntity),
      };
    },
    async update({ post }) {
      await db.post.update({
        where: {
          id: post.id,
        },
        data: {
          createdAt: post.createdAt,
          published: post.published,
          title: post.title,
        },
      });
    },
  };

  function toEntity(post: PostModel) {
    return new Post({
      id: post.id,
      createdAt: post.createdAt,
      published: post.published,
      title: post.title,
    });
  }
}
