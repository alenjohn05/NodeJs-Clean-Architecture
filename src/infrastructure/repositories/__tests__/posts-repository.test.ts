import { mockDeep } from 'jest-mock-extended';
import { PrismaClient } from '@prisma/client';
import { makePostsRepository } from '../posts-repository';
import { Post } from '@domain/entities';

describe('postsRepository', () => {
  function setup() {
    const db = mockDeep<PrismaClient>();

    const postsRepository = makePostsRepository({
      db,
    });

    return {
      db,
      postsRepository,
    };
  }

  describe('create', () => {
    test('creates post and returns the id', async () => {
      const { db, postsRepository } = setup();

      db.post.create.mockResolvedValue({
        id: 1,
        createdAt: new Date(2022, 1, 1),
        published: true,
        title: 'Mock title',
      });

      const result = await postsRepository.create({
        post: new Post({
          createdAt: new Date(2022, 1, 1),
          published: true,
          title: 'Mock title',
        }),
      });

      expect(result.id).toEqual(1);
      expect(db.post.create).toHaveBeenCalledTimes(1);
      expect(db.post.create.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    test('deletes post', async () => {
      const { db, postsRepository } = setup();

      await postsRepository.delete({ id: 1 });

      expect(db.post.delete).toBeCalledTimes(1);
      expect(db.post.delete.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('getById', () => {
    test('returns null if the post does not exist', async () => {
      const { db, postsRepository } = setup();

      const result = await postsRepository.getById({ id: 1 });

      expect(result).toBeNull();
      expect(db.post.findFirst).toHaveBeenCalledTimes(1);
      expect(db.post.findFirst.mock.calls[0]).toMatchSnapshot();
    });

    test('returns post if it exists', async () => {
      const { db, postsRepository } = setup();

      db.post.findFirst.mockResolvedValue({
        id: 1,
        createdAt: new Date(2022, 1, 1),
        published: true,
        title: 'Mock title',
      });

      const result = await postsRepository.getById({ id: 1 });

      expect(result).toMatchSnapshot();
    });
  });

  describe('list', () => {
    test('list posts', async () => {
      const { db, postsRepository } = setup();

      db.post.count.mockResolvedValue(2);

      db.post.findMany.mockResolvedValue([
        {
          id: 1,
          createdAt: new Date(2022, 1, 1),
          published: true,
          title: 'Mock post one',
        },
        {
          id: 2,
          createdAt: new Date(2022, 2, 1),
          published: false,
          title: 'Mock post two',
        },
      ]);

      const result = await postsRepository.list({ pageNumber: 1, pageSize: 10 });

      expect(db.post.findMany).toBeCalledTimes(1);
      expect(db.post.findMany.mock.calls[0]).toMatchSnapshot();
      expect(result).toMatchSnapshot();
    });
  });

  describe('update', () => {
    test('updates post', async () => {
      const { db, postsRepository } = setup();

      await postsRepository.update({
        post: new Post({
          id: 1,
          createdAt: new Date(2022, 1, 1),
          published: true,
          title: 'Mock title',
        }),
      });

      expect(db.post.update).toHaveBeenCalledTimes(1);
      expect(db.post.update.mock.calls[0]).toMatchSnapshot();
    });
  });
});
