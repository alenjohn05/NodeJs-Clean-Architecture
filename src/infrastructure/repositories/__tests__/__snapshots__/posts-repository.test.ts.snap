// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postsRepository create creates post and returns the id 1`] = `
[
  {
    "data": {
      "createdAt": 2022-02-01T00:00:00.000Z,
      "published": true,
      "title": "Mock title",
    },
  },
]
`;

exports[`postsRepository delete deletes post 1`] = `
[
  {
    "where": {
      "id": 1,
    },
  },
]
`;

exports[`postsRepository getById returns null if the post does not exist 1`] = `
[
  {
    "where": {
      "id": 1,
    },
  },
]
`;

exports[`postsRepository getById returns post if it exists 1`] = `
Post {
  "createdAt": 2022-02-01T00:00:00.000Z,
  "id": 1,
  "published": true,
  "title": "Mock title",
}
`;

exports[`postsRepository list list posts 1`] = `
[
  {
    "skip": 0,
    "take": 10,
  },
]
`;

exports[`postsRepository list list posts 2`] = `
{
  "count": 2,
  "posts": [
    Post {
      "createdAt": 2022-02-01T00:00:00.000Z,
      "id": 1,
      "published": true,
      "title": "Mock post one",
    },
    Post {
      "createdAt": 2022-03-01T00:00:00.000Z,
      "id": 2,
      "published": false,
      "title": "Mock post two",
    },
  ],
}
`;

exports[`postsRepository update updates post 1`] = `
[
  {
    "data": {
      "createdAt": 2022-02-01T00:00:00.000Z,
      "published": true,
      "title": "Mock title",
    },
    "where": {
      "id": 1,
    },
  },
]
`;
